src/common/file_stream.h.bac

```diff
@@ -0,0 +1,201 @@
+      //ABORT_IF(unlink(name_.c_str()), "Error while deleting '{}'", name_);
+    }
+    // if(fd_ != -1 && close(fd_)) {
+    //   ABORT("Could not close file {}", fd_ );
+    // }
+  }
+
+  void seek(size_t pos) {
+    lseek(fd_, pos, SEEK_SET);
+  }
+
+  int getFileDescriptor() { return fd_; }
+
+  std::streambuf* getBuffer() { buffer_.get(); }
+
+  std::string getFileName() { return name_; }
+};
+
+class InputFileStream {
+private:
+  std::unique_ptr<std::istream> istream_;
+  marian::filesystem::Path file_;
+
+public:
+  InputFileStream(const std::string& file)
+  : file_(file) {
+    ABORT_IF(!marian::filesystem::exists(file_),"File '{}' does not exist", file);
+
+    if(file_.extension() == marian::filesystem::Path(std::string(".gz")))
+      istream_.reset(new zstr::ifstream(file_));
+    else
+      istream_.reset(new std::ifstream(file_));
+  }
+
+  InputFileStream(TemporaryFile& tempfile) {
+    tempfile.seek(0);
+    istream_.reset(new std::istream(tempfile.getBuffer()));
+  }
+
+  InputFileStream(std::istream& strm) {
+    istream_.reset(new std::istream(strm.rdbuf()));
+  }
+
+
+  operator std::istream&() { return *istream_; }
+
+  operator bool() { return (bool)*istream_; }
+
+  template <typename T>
+  friend InputFileStream& operator>>(InputFileStream& stream, T& t) {
+    *(stream.istream_) >> t;
+    return stream;
+  }
+
+  template <typename T>
+  size_t read(T* ptr, size_t num = 1) {
+    istream_->read((char*)ptr, num * sizeof(T));
+    return num * sizeof(T);
+  }
+
+  std::string path() { return file_.string(); }
+
+  bool empty() { return istream_->peek() == std::ifstream::traits_type::eof(); }
+
+};
+
+class OutputFileStream {
+private:
+  std::unique_ptr<std::ostream> ostream_;
+  marian::filesystem::Path file_;
+
+public:
+  OutputFileStream(const std::string& file) : file_(file) {
+    if(file_.extension() == marian::filesystem::Path(std::string(".gz")))
+      ostream_.reset(new zstr::ofstream(file_));
+    else
+      ostream_.reset(new std::ofstream(file_));
+  }
+
+  OutputFileStream(TemporaryFile& tempfile) {
+    tempfile.seek(0);
+    ostream_.reset(new std::ostream(tempfile.getBuffer()));
+  }
+
+  OutputFileStream(std::ostream& strm) {
+    ostream_.reset(new std::ostream(strm.rdbuf()));
+  }
+
+  operator std::ostream&() { return *ostream_; }
+
+  operator bool() { return (bool)*ostream_; }
+
+  template <typename T>
+  friend OutputFileStream& operator<<(OutputFileStream& stream, const T& t) {
+    *(stream.ostream_) << t;
+    return stream;
+  }
+
+  // handle things like std::endl which is actually a function not a value
+  friend OutputFileStream& operator<<(OutputFileStream& stream, std::ostream& (*var)(std::ostream&)) {
+    *(stream.ostream_) << var;
+    return stream;
+  }
+
+  template <typename T>
+  size_t write(const T* ptr, size_t num = 1) {
+    ostream_->write((char*)ptr, num * sizeof(T));
+    return num * sizeof(T);
+  }
+
+  std::string path() { return file_.string(); }
+};
```

Comment:
Do we need to add the unused file into master? I think we should keep it in this branch only. Maybe the same with zstr.